docs/usage.md

```diff
@@ -0,0 +1,236 @@
+
+The operating system shouldn't matter, but it should be consistent across all benchmarks, and in the past we used the latest stable version of Debian Linux.
+
+Note that we're deliberately not using distributed execution, but running k6 from a single node at a time. The goal of the benchmarks is not to load test the SUT, but to measure k6 performance on a single machine.
+
+
+## OS performance tweaks
+
+Before running the benchmarks it's important that some operating system configuration changes are done on the load generator machines to ensure k6 can fully utilize all system resources. This includes increasing the network connection limit and local port range, disabling swap space, among others.
+
+Please consult the [OS fine-tuning section of the "Running large tests"](https://k6.io/docs/testing-guides/running-large-tests/#os-fine-tuning) article for details.
+
+
+## Test scripts
+
+We currently have 3 k6 [scripts](/scripts) we will run:
+
+- `website.js`: a general purpose script that replicates common real-world scenarios: loading the front page and static assets, and logging in using some predefined credentials. It uses popular k6 features, such as custom metrics, checks, thresholds and `sleep()`.
+  The purpose of this script is not to achieve the highest RPS, but to track general performance.
+
+- `RPS-optimized.js`: a simple script that tries to achieve the highest possible RPS.
+
+- `file-upload.js`: a script that uploads a large file, used to track memory usage and network throughput.
+
+
+## Helper scripts
+
+We use a couple of scripts to help us with running the benchmarks and compiling the results:
+
+- [k6bench.sh](/k6bench.sh): this is a wrapper script around `k6 run` that starts k6 and the metric collection processes in the background.
+  It redirects the k6 output to a file, and outputs metric results to stdout in CSV format every 5 seconds by default.
+
+- [k6bench.gnuplot](/k6bench.gnuplot): this is a [gnuplot](http://www.gnuplot.info/) script that generates a graph from the CSV results created by `k6bench.sh`.
+  It doesn't create the prettiest graphs, but it was chosen because it's scriptable, so it can be easily automated.
+
+
+## Dependencies
+
+The benchmarks depend on a few utilities to collect the performance results:
+
+- [`smem`](https://www.selenic.com/smem/) for RAM usage, specifically proportional set size (PSS).
+
+  As it turns out, it's rather tricky to determine the precise physical RAM usage of a single process in Linux. Tools like `ps` and `htop` report the resident set size (RSS), which includes the size of any shared libraries the process may be using, and is not an accurate value of how much physical memory the process is using. The PSS value is more precise, but is more difficult to measure. `smem` is seemingly the only tool that simplifies tracking this, but it's unfortunately written in Python, so its own performance is not great, and it can use a lot of CPU and take some time to generate the result.
```

Comment:
Unrelated to the article:

> which includes the size of any shared libraries the process may be using

k6 should in practice has only libc as the shared library. So I am not certain if using something that has performance problems is worth it.

Also, if it uses memory - it uses memory. On most load generator machines, k6 *will* be the only process or one of very few, so not counting shared libraries seems like a strange choice.

I doubt that the difference is ... relevant to the performance testing either way. But if `smem` is problematic it might be worth it just removing it and using `ps`.

So maybe we should open an issue?

Reply:
@mstoykov It's not just about shared memory. RSS is an inaccurate measurement in general. If you start `k6bench.sh`, and track the RSS value for the k6 process, you'll see that it will vary in consistency, and sometimes be ~2-10 MB higher than what `smem` reports. Whether that's only due to libc, or some other libraries, or just general inaccuracies with RSS, I'm not entirely sure. Since we do want to be as accurate as possible in benchmarks, getting this right seems important. Especially since "memory usage" will be extrapolated by users of other operating systems, and we don't want to report values that rely on some Linux quirks.

The inaccuracy of RSS is mentioned in the [`proc(5)` manpage](https://man7.org/linux/man-pages/man5/proc.5.html), and summarized well in [this comment](https://github.com/oshi/oshi/issues/1516#issuecomment-773588484).

Note specifically this point:
> Some of these values are inaccurate because of a kernel-
> internal scalability optimization.  If accurate values are
> required, use /proc/pid/smaps or /proc/pid/smaps_rollup
> instead, which are much slower but provide accurate,
> detailed information.

[This is essentially what `smem` does.](https://github.com/kwkroeger/smem/blob/master/smem#L107-L109) So if anything, the slowness of `smem` is due to Linux, and not because of Python itself, so we should probably remove the slight jab at Python here.

Besides, the performance is not terrible to be a concern. We currently output in 5s intervals, which is plenty for `smem` to run, and for the script to gather other statistics.

So I'd rather we default to more accurate measurements, and leave this as is.